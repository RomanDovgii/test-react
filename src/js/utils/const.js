export const MAXIMUM_USERS_PER_PAGE = 50;

export const AppRoute = {
  SELECT: `/select`,
  TABLE: `/table`
};
