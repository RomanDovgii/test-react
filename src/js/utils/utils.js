export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const sortToTop = (a, b) => {
  switch (true) {
    case a > b:
      return -1;
    case a < b:
      return 1;
    default:
      return 0;
  }
};

export const sortToBot = (a, b) => {
  switch (true) {
    case a < b:
      return -1;
    case a > b:
      return 1;
    default:
      return 0;
  }
};
