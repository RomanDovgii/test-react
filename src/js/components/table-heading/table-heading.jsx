import React, {createRef} from "react";
import {connect} from "react-redux";
import {sortUsers, updateCurrentUsers} from "../../store/action/action";
import {FilterDirection, FilterType} from "../../utils/const";
import {tableHeadingType} from "../../types/types";

const TableHeading = ({onFilterClick}) => {
  const idRef = createRef();
  const firstNameRef = createRef();
  const lastNameRef = createRef();
  const emailRef = createRef();
  const phoneRef = createRef();

  return (
    <tr className="table__row">
      <th className="table__heading ">
        <button
          className="table__button"
          ref={idRef}
          onClick={() => {
            const refs = [idRef.current, firstNameRef.current, lastNameRef.current, emailRef.current, phoneRef.current];
            onFilterClick(idRef.current, refs);
          }}
        >
          id
        </button>
      </th>
      <th className="table__heading">
        <button
          className="table__button"
          ref={firstNameRef}
          onClick={() => {
            const refs = [idRef.current, firstNameRef.current, lastNameRef.current, emailRef.current, phoneRef.current];
            onFilterClick(firstNameRef.current, refs);
          }}
        >
          firstName
        </button>
      </th>
      <th className="table__heading">
        <button
          className="table__button"
          ref={lastNameRef}
          onClick={() => {
            const refs = [idRef.current, firstNameRef.current, lastNameRef.current, emailRef.current, phoneRef.current];
            onFilterClick(lastNameRef.current, refs);
          }}
        >
          lastName
        </button>
      </th>
      <th className="table__heading">
        <button
          className="table__button"
          ref={emailRef}
          onClick={() => {
            const refs = [idRef.current, firstNameRef.current, lastNameRef.current, emailRef.current, phoneRef.current];
            onFilterClick(emailRef.current, refs);
          }}
        >
          email
        </button>
      </th>
      <th className="table__heading">
        <button
          className="table__button"
          ref={phoneRef}
          onClick={() => {
            const refs = [idRef.current, firstNameRef.current, lastNameRef.current, emailRef.current, phoneRef.current];
            onFilterClick(phoneRef.current, refs);
          }}
        >
          phone
        </button>
      </th>
    </tr>
  );
};

TableHeading.propTypes = tableHeadingType;

const mapDispatchToPorps = (dispatch) => ({
  onFilterClick(ref, refs) {
    let filterName = ref.textContent;
    let filterDirection;

    refs.forEach((elementRef) => {
      if (elementRef !== ref) {
        elementRef.classList.remove(`table__button--bot-to-top`);
        elementRef.classList.remove(`table__button--top-to-bot`);
      }
    });

    switch (true) {
      case ref.classList.contains(`table__button--top-to-bot`):
        filterDirection = FilterDirection.TO_TOP;
        ref.classList.remove(`table__button--top-to-bot`);
        ref.classList.add(`table__button--bot-to-top`);
        break;
      case ref.classList.contains(`table__button--bot-to-top`):
        filterName = FilterType.NONE;
        ref.classList.remove(`table__button--bot-to-top`);
        break;
      default:
        filterDirection = FilterDirection.TO_BOT;
        ref.classList.add(`table__button--top-to-bot`);
        break;
    }

    const filterInfo = {
      filter: filterName,
      direction: filterDirection,
    };
    dispatch(sortUsers(filterInfo));
    dispatch(updateCurrentUsers());
  }
});

export {TableHeading};
export default connect(null, mapDispatchToPorps)(TableHeading);
