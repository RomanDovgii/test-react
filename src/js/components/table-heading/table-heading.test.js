import React from "react";
import renderer from "react-test-renderer";
import {TableHeading} from "./table-heading";
import {noop} from "../../utils/utils";

it(
    `Should render TableHeading`,
    () => {
      const tree = renderer
      .create(
          <TableHeading
            onFilterClick={noop}
          />
      )
      .toJSON();

      expect(tree).toMatchSnapshot();
    }
);
