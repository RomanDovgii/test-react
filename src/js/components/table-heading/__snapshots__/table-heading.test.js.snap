// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render TableHeading 1`] = `
<tr
  className="table__row"
>
  <th
    className="table__heading "
  >
    <button
      className="table__button"
      onClick={[Function]}
    >
      id
    </button>
  </th>
  <th
    className="table__heading"
  >
    <button
      className="table__button"
      onClick={[Function]}
    >
      firstName
    </button>
  </th>
  <th
    className="table__heading"
  >
    <button
      className="table__button"
      onClick={[Function]}
    >
      lastName
    </button>
  </th>
  <th
    className="table__heading"
  >
    <button
      className="table__button"
      onClick={[Function]}
    >
      email
    </button>
  </th>
  <th
    className="table__heading"
  >
    <button
      className="table__button"
      onClick={[Function]}
    >
      phone
    </button>
  </th>
</tr>
`;
